File: haskell/src/parser.hs
module Parser where

import Interpreter
import Data.List

type Parser = [Program]

push :: Parser -> Parser
push p = []:p

pop :: Parser -> Parser
pop (p:p':ps) = ((Loop $ reverse p):p'):ps

-- Add an instruction to the top of 
add :: Instruction -> Parser -> Parser
add i (p:ps) = (i:p):ps

parse :: String -> Program
parse src = reverse . head $ foldl' (flip parseChar) [[]] src

parseChar :: Char -> Parser -> Parser
parseChar '.' = add Put
parseChar ',' = add Get
parseChar '>' = add Next
parseChar '<' = add Prev
parseChar '+' = add $ Add 1
parseChar '-' = add $ Add (-1)
parseChar '[' = push
parseChar ']' = pop
parseChar _   = id

File: scala/src/main/scala/com/hardlymundane/bf/Parser.scala
package com.hardlymundane.bf

import com.hardlymundane.bf.AST._

import scala.annotation.tailrec

object Parser {

  type Stack = List[Program]

  private def push(s: Stack, i: Instruction): Stack = (i :: s.head) :: s.tail

  private def pushFrame(s: Stack): Stack = Nil :: s

  def parse(src: String): Program = {
    @tailrec
    def run(stack: Stack, src: List[Char]): Program = src match {
      case Nil => stack.head.reverse
      case head :: tail => head match {
        case '+' => run(push(stack, Add(1)), tail)
        case '-' => run(push(stack, Add(-1)), tail)
        case '<' => run(push(stack, Prev), tail)
        case '>' => run(push(stack, Next), tail)
        case '.' => run(push(stack, Print), tail)
        case ',' => run(push(stack, Read), tail)
        case '[' => run(pushFrame(stack), tail)
        case ']' => run(push(stack.tail, Loop(stack.head.reverse)), tail)
        case _   => run(stack, tail)
      }
    }

    run(List(Nil), src.toList)
  }
}
